units/migrate_volume_job.go

```diff
@@ -0,0 +1,298 @@
+	stopJob, foundJob := j.env.RemoteQueue().Get(ctx, j.SpawnhostStopJobID)
+	if !foundJob {
+		j.AddRetryableError(errors.Errorf("spawnhost-stop job '%s' could not be found", j.SpawnhostStopJobID))
+		return false, true
+	}
+
+	if stopJob.Status().Completed {
+		j.AddError(errors.Errorf("host '%s' failed to stop", j.SpawnhostStopJobID))
+		return false, true
+	}
+
+	return false, false
+}
+
+// startNewHost attempts to start a new host with the volume attached
+// It returns a boolean indicating if an error was added to the job
+func (j *volumeMigrationJob) startNewHost(ctx context.Context) bool {
+	// Ensure volume has been detached
+	if j.volume.Host == j.InitialHostID {
+		j.AddRetryableError(errors.New("volume still attached to host"))
+		return true
+	}
+
+	j.ModifyOptions.HomeVolumeID = j.VolumeID
+	intentHost, err := cloud.CreateSpawnHost(ctx, j.ModifyOptions, j.env.Settings())
+	if err != nil {
+		j.AddError(errors.Wrapf(err, "creating new intent host"))
+		return true
+	}
+
+	if err := intentHost.Insert(); err != nil {
+		j.AddError(errors.Wrap(err, "inserting new intent host"))
+		return true
+	}
+	grip.Info(message.Fields{
+		"message":  "new intent host created",
+		"attempts": j.RetryInfo().CurrentAttempt,
+	})
+	j.NewHostID = intentHost.Id
+	return false
+}
+
+// finishJob marks the job as completed and attempts some additional cleanup if this is the job's final attempt
+func (j *volumeMigrationJob) finishJob(ctx context.Context) {
+	// Amboy's ShouldRetry() method does not account for reaching MaxAttempts, so manually determine if the current attempt is the last
+	retryInfo := j.RetryInfo()
+	isFinalAttempt := retryInfo.CurrentAttempt == (retryInfo.MaxAttempts - 1)
+	if !retryInfo.ShouldRetry() || isFinalAttempt {
+		// If a new host was never created, restart initial host with volume attached
+		if j.NewHostID == "" {
+			if hasErr := j.restartInitialHost(ctx); hasErr {
+				return
+			}
+		}
+
+		if err := j.volume.SetMigrating(false); err != nil {
+			j.AddRetryableError(err)
+			return
+		}
+
+	}
+
+	j.MarkComplete()
+}
+
+// restartInitialHost restarts the stopped host with the volume attached
+// This is a fallback so that we can attempt to leave users with a usable workstation in case other operations fail
+// It returns a boolean indicating whether an error was added to the job
+func (j *volumeMigrationJob) restartInitialHost(ctx context.Context) bool {
+	if j.volume.Host == "" {
+		_, err := cloud.AttachVolume(ctx, j.VolumeID, j.InitialHostID)
+		if err != nil {
+			j.AddError(errors.Wrapf(err, "reattaching volume to initial host"))
+			return true
+
+		}
+	}
+
+	ts := utility.RoundPartOfMinute(1).Format(TSFormat)
+	if err := j.env.RemoteQueue().Put(ctx, NewSpawnhostStartJob(j.initialHost, evergreen.User, ts)); err != nil {
+		j.AddError(errors.Wrap(err, "enqueuing job to restart initial host"))
+		return true
+	}
+	return false
+
+}
+
+func (j *volumeMigrationJob) getVolume() (*host.Volume, error) {
```

Comment:
This helper function ultimately only moves one if statement out of the main loop so I'm not certain it's really needed